### PR TITLE
Preserve all node events

### PR DESCRIPTION
**Pull request by mpolden**
Node events are now limited by a total size limit, instead of one per type.

@bratseth

### CODE REVIEW

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/node/History.java [hunk 1]
```diff
@@ -1,85 +1,83 @@
...
 
-    /** Returns this event if it is present in this history */
-    public Optional<Event> event(Event.Type type) { return Optional.ofNullable(events.get(type)); }
+    /** Returns the latest event of given type, if it is present in this history */
+    public Optional<Event> event(Event.Type type) {
+        return events.stream().filter(event -> event.type() == type).max(Comparator.comparing(Event::at));
+    }
 
     /** Returns true if a given event is registered in this history at the given time */
     public boolean hasEventAt(Event.Type type, Instant time) {
```
Comment (bratseth) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/node/History.java:47]: Change this to hasLastEventAt?

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/node/History.java [hunk 2]
```diff
@@ -1,85 +1,83 @@
...
     }
 
-    /** Returns this event if it is present in this history */
-    public Optional<Event> event(Event.Type type) { return Optional.ofNullable(events.get(type)); }
+    /** Returns the latest event of given type, if it is present in this history */
+    public Optional<Event> event(Event.Type type) {
+        return events.stream().filter(event -> event.type() == type).max(Comparator.comparing(Event::at));
+    }
 
     /** Returns true if a given event is registered in this history at the given time */
```
Comment (bratseth) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/node/History.java:46]:
```suggestion
    /** Returns true if the last event of this type is registered in this history at the given time */
```

**Review by bratseth — APPROVED** [review]

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/node/History.java [hunk 3]
```diff
@@ -1,85 +1,83 @@
...
-
-    private ImmutableMap.Builder<Event.Type, Event> builderWithout(Event.Type type) {
-        ImmutableMap.Builder<Event.Type, Event> builder = new ImmutableMap.Builder<>();
-        for (Event event : events.values())
-            if (event.type() != type)
-                builder.put(event.type(), event);
-        return builder;
+        List<Event> copy = new ArrayList<>(events);
+        if (!copy.isEmpty()) {
+            // Let given event overwrite the latest if they're of the same type. Some events may be repeated, such as
```
[the remark this answers is not in the document]
Comment (mpolden) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/node/History.java:72]: Not sure I agree. If some other code misbehaves and adds the same event multiple times, do we want that to push all other events out of the history?

I suppose we could add a counter to the event if knowing about repeating ones is important.

[the remark this answers is not in the document]
Comment (bratseth) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/node/History.java:72]: We could have some code that added a pair of events and pushed everything out too, should we protect against that as well? :-)

"The last event of each type" and "the last 15 events" are simple and well-defined. Future devs will have some simple model like that in their head while with this real behavior or more complicated, is what I mean. Anyway, feel free to merge if you prefer, we can address it when it becomes a problem.

**Comment by freva** [comment]
Wont this result in fewer "interesting" events? With "up" as a separate state, 8 upgrade cycles (2 weeks) will clear all other events. Even if we add some special handling for "up" and "down", I'm much more interested in when a host was provisioned than its 10th last reboot.

**Comment by mpolden** [comment]
Yes, but at the same time it's useful to have a chronological log of events (this was an action item from a post mortem).

If we can define "interesting", perhaps we can truncate the history in a smarter way. E.g. always keep the first X and last Y events?

**Comment by freva** [comment]
How about limit to 15 events, but keep at least the last event of each type. The remaining spots can be taken by the latest chronological events?

**Comment by mpolden** [comment]
Then we would need a higher limit because there's currently 21 distinct event types.

**Comment by freva** [comment]
In practice I don't think this is an issue. I checked all nodes in all systems, here's distribution of history length:
Hosts: {1: 16, 3: 48, 4: 48, 5: 649, 6: 22, 7: 140, 8: 75, 9: 637, 10: 484, 11: 182, 12: 22, 13: 7, 14: 2, 15: 1}
Nodes: {1: 47, 2: 27, 3: 6947, 4: 1206, 5: 257, 6: 183, 7: 141, 8: 26, 9: 13, 10: 3}

**Comment by mpolden** [comment]
I think this must be chronological in order to be useful for debugging. A history that is sometimes chronological and sometimes not is even more confusing than the current state.

However, we also have code making decisions based on the presence of, or elapsed time since, an event. Maybe it's best to have the chronological history as a separate field? Then have a toggle between "last by type" and "chronological" in the console.

**Pull request merged**